### PR TITLE
confdgnmi - initial version (0.0.1) version of ConfD gNMI Adapter

### PR DESCRIPTION
**Pull request by micnovak**
Initial version of ConfD gNMI Adapter

### CODE REVIEW

File: confdgnmi/Makefile [hunk 1]
```diff
@@ -0,0 +1,91 @@
+######################################################################
+# Interface Status example
+# (C) 2006-2009 Tail-f Systems
```
Comment (jomiklos) [inline confdgnmi/Makefile]: wrong year(s)?

File: confdgnmi/Makefile [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+
+######################################################################
+# Example specific definitions and rules
+
+CONFD_FLAGS ?= --addloadpath $(CONFD_DIR)/etc/confd
+#EXTRA_LINK_FLAGS +=-F rt:router-id
+START_FLAGS ?=
+SRC_DIR=src
+PCG_DIR=$(SRC_DIR)
```
Comment (jomiklos) [inline confdgnmi/Makefile:35]: PCG? :) plus, seems to be just unnecessary clone of SRC_DIR...

File: confdgnmi/Makefile [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+
+#all:	ietf-ip.fxs ietf-interfaces.fxs ietf-ipv6-unicast-routing.fxs \
```
Comment (jomiklos) [inline confdgnmi/Makefile]: remove leftovers?

File: confdgnmi/README.md [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+We use (part) of `ietf-interfaces.yang` as sample datamodel.
+
+### gNMI interface 
+
+#### Capabilities
+
+`rpc Capabilities(CapabilityRequest) returns (CapabilityResponse);`
+
+`Capability` returns list of `supported_models` and list of `encodings`.
+It can also return lis of gNMI extensions.
```
Comment (jomiklos) [inline confdgnmi/README.md]: lisT

File: confdgnmi/README.md [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+It can also return lis of gNMI extensions.  
+
+Each model has following attributes: 
+ 
+`name` (string), `organization` (string), `version`
+
+##### Implementation 
+
+
+Read values in  `"/ncm:netconf-state/ncm:capabilities/ncm:capability"`
```
Comment (jomiklos) [inline confdgnmi/README.md]: me? now? seems to be too cryptic, what this means, what is the step for, etc.

File: confdgnmi/README.md [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+`update`, each with value.  
+
+* Since there is a path associated with each value, for requests on nested lists
+we have to find out, which elements are keys and add them to the Path(s).
+
+* What value type should be for empty or presence container ??? (bool ???)
+
+##### Implementation 
+
+Use `maapi_save_config`  - can be used even for operational data (` MAAPI_CONFIG_WITH_OPER`).
```
Comment (jomiklos) [inline confdgnmi/README.md]: use use in single sentence, only one use needed? :)

File: confdgnmi/README.md [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+#### Set
+  
+`rpc Set(SetRequest) returns (SetResponse);`
+
+Pass in request list of paths to `delete`, list of paths and values to 
+be replaced (`replace`) and list of paths and values to be updated (`update`).
+
+TODOs:
+
+* difference between replace and update.   -> 3.4.4.
```
Comment (jomiklos) [inline confdgnmi/README.md]: what is 3.4.4.? reference to some rfc/document?

File: confdgnmi/README.md [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+
+Each gNMI `Set` call should be treated as transaction
+ 
+#### Subscribe
+  
+`rpc Subscribe(stream SubscribeRequest) returns (stream SubscribeResponse);`
+
+Stream subscription requests and get stream of subscription responses.
+
+In some cases, subscription response is similar to `Get`, this means all data in
```
Comment (jomiklos) [inline confdgnmi/README.md]: cases of what? is a bit confusing - maybe rephrase /clarify?

File: confdgnmi/datagen.py [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+        <type xmlns:ianaift="urn:ietf:params:xml:ns:yang:iana-if-type">
+            gigabitEthernet
+        </type>             
+    </interface>
+    """.format(name)
+    return interface
+
+
+def ifname(index):
+    return "cp_ont_{}".format(index)
```
Comment (jomiklos) [inline confdgnmi/datagen.py]: cryptic - what is this shortcut/representation of? :)

File: confdgnmi/src/confd_gnmi_adapter.py [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+    class CapabilityModel:
+        name: str
+        organization: str
+        version: str
+        schema: str = ""
+
+    @abstractmethod
+    def capabilities(self):
+        """
+        Invoke capabilities oon adapter and return list of Capabilities
```
Comment (jomiklos) [inline confdgnmi/src/confd_gnmi_adapter.py]: oon = on?

File: confdgnmi/src/confd_gnmi_api_adapter.py [hunk 11]
```diff
@@ -0,0 +1,348 @@
...
+        self.addr: str = ""
+        self.port: int = 0
+        self.username: str = ""
+        self.password: str = ""
+        self.maapisock = None
+        self.mp_inst = None
+
+    @staticmethod
+    def set_confd_debug_level(level):
+        if level == "debug":
```
Comment (jomiklos) [inline confdgnmi/src/confd_gnmi_api_adapter.py:31]: nice const hashmap/dict candidate...

File: confdgnmi/src/confd_gnmi_api_adapter.py [hunk 12]
```diff
@@ -0,0 +1,348 @@
...
+        elif level == "trace":
+            GnmiConfDApiServerAdapter.confd_debug_level = _confd.TRACE
+        elif level == "proto":
+            GnmiConfDApiServerAdapter.confd_debug_level = _confd.PROTO_TRACE
+        elif level == "silent":
+            GnmiConfDApiServerAdapter.confd_debug_level = _confd.SILENT
+        else:
+            log.warning("Unknow confd debug level %s", level)
+
+    @staticmethod
```
Comment (jomiklos) [inline confdgnmi/src/confd_gnmi_api_adapter.py:42]: feels a bit strange to modify class props instead of instance, any special reason for not making these instance parameters? or constructor params...

File: confdgnmi/src/confd_gnmi_api_adapter.py [hunk 13]
```diff
@@ -0,0 +1,348 @@
...
+                root = maagic.get_root(t)
+                values = []
+                count = 0
+                # format
+                # http://tail-f.com/ns/confd-progress?module=tailf-confd-progress&revision=2020-06-29
+                for val in root.netconf_state.capabilities.capability:
+                    log.debug("val=%s", val)
+                    if "module=" in val:
+                        el = val.split("?")
+                        if len(el) > 1:
```
Comment (jomiklos) [inline confdgnmi/src/confd_gnmi_api_adapter.py:139]: use regexp on-liner to simplify & extract capability? something like:

```python
cap_name = re.search(r"module=(.*)&", val).group()
```

File: confdgnmi/src/confd_gnmi_api_adapter.py [hunk 14]
```diff
@@ -0,0 +1,348 @@
...
+    #                                       ietf_netconf_monitoring_ns.ns.hash,
+    #                                       "/ncm:get-schema")
+    #     log.debug("values=%s", values)
+    #     maapi_low.end_user_session(self.maapisock)
+    #     log.info("<== get_schema_action")
+
+    def capabilities(self):
+        log.info("==>")
+        ns_list = self.get_netconf_capabilities()
+        # ns_list = _confd.get_nslist()
```
Comment (jomiklos) [inline confdgnmi/src/confd_gnmi_api_adapter.py]: commented lines - leftovers?

File: confdgnmi/src/confd_gnmi_api_adapter.py [hunk 15]
```diff
@@ -0,0 +1,348 @@
...
+            # maapi_low.finish_trans(self.maapisock, tid)
+            # maapi_low.end_user_session(self.maapisock)
+        except Exception as e:
+            log.exception(e)
+
+        log.debug("<== save_str=%s", save_str)
+        return save_str
+
+    def get_xml_leaves(self, elem):
+        leaves = []
```
Comment (jomiklos) [inline confdgnmi/src/confd_gnmi_api_adapter.py:212]: shorten/simplify?

```python
def get_xml_leaves(self, elem):
    if len(elem) <= 0:
        return [elem];
    return [self.get_xml_leaves(e) for e in elem];
```

(quick glance, untested...)

fn naming is however a bit confusing though, as it seems to return list of one-element lists?

File: confdgnmi/src/confd_gnmi_common.py [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+
+def common_optparse_options(parser):
+    parser.add_option("--logging", action="store", dest="logging",
+                      help="Logging level [error, warning, info, debug]",
+                      default="warning")
+
+
+def common_optparse_process(opt, log):
+    level = None
+    if opt.logging == "error":
```
Comment (jomiklos) [inline confdgnmi/src/confd_gnmi_common.py:23]: const dict candidate...

**Review by jomiklos — COMMENTED** [review]
quick glance through.
I still need to look into high level architecture of classes/files, which is bit cumbersome on github page directly...

File: confdgnmi/Makefile (hunk 2 above)
Comment (micnovak) [reply confdgnmi/Makefile:35]: This is on purpose for case we want to move python (packages) to other location.

File: confdgnmi/README.md (hunk 5 above)
Comment (micnovak) [reply confdgnmi/README.md]: Added reference to `ietf-netconf-monitoring.yang`

File: confdgnmi/Makefile (hunk 1 above)
Comment (micnovak) [reply confdgnmi/Makefile]: fixed

File: confdgnmi/Makefile (hunk 3 above)
Comment (micnovak) [reply confdgnmi/Makefile]: removed

File: confdgnmi/README.md (hunk 4 above)
Comment (micnovak) [reply confdgnmi/README.md]: fixed

File: confdgnmi/README.md (hunk 6 above)
Comment (micnovak) [reply confdgnmi/README.md]: first removed

File: confdgnmi/README.md (hunk 7 above)
Comment (micnovak) [reply confdgnmi/README.md]: added gNMI spec.

File: confdgnmi/README.md (hunk 8 above)
Comment (micnovak) [reply confdgnmi/README.md]: removed

File: confdgnmi/datagen.py (hunk 9 above)
Comment (micnovak) [reply confdgnmi/datagen.py]: changed to `if_`

File: confdgnmi/src/confd_gnmi_adapter.py (hunk 10 above)
Comment (micnovak) [reply confdgnmi/src/confd_gnmi_adapter.py]: fixed

File: confdgnmi/Makefile (hunk 2 above)
Comment (jomiklos) [reply confdgnmi/Makefile:35]: what i mean is it's cryptic, i do not understand what does the shortcut mean...

File: confdgnmi/src/confd_gnmi_api_adapter.py (hunk 11 above)
Comment (micnovak) [reply confdgnmi/src/confd_gnmi_api_adapter.py:31]: not resolved for now

File: confdgnmi/src/confd_gnmi_api_adapter.py (hunk 12 above)
Comment (micnovak) [reply confdgnmi/src/confd_gnmi_api_adapter.py:42]: only one confd level is allowed for now (by command line options), but added to `connect` method so it can be changed for the instance, the class variables are mostly holders for default values.

File: confdgnmi/src/confd_gnmi_api_adapter.py (hunk 13 above)
Comment (micnovak) [reply confdgnmi/src/confd_gnmi_api_adapter.py:139]: Not fixed for now

File: confdgnmi/src/confd_gnmi_api_adapter.py (hunk 14 above)
Comment (micnovak) [reply confdgnmi/src/confd_gnmi_api_adapter.py]: yes, still do not know what to do with other possible implementation

File: confdgnmi/src/confd_gnmi_api_adapter.py (hunk 15 above)
Comment (micnovak) [reply confdgnmi/src/confd_gnmi_api_adapter.py:212]: should return leaf elements of XML tree

File: confdgnmi/src/confd_gnmi_common.py (hunk 16 above)
Comment (micnovak) [reply confdgnmi/src/confd_gnmi_common.py:23]: not fixed for now

File: confdgnmi/Makefile [hunk 17]
```diff
@@ -0,0 +1,87 @@
+######################################################################
```
Comment (cohult) [inline confdgnmi/Makefile:1]: The copyright header should be removed as this is an open-source demo.

File: confdgnmi/Makefile [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+PCG_DIR=$(SRC_DIR)
+
+
+all: gnmi_proto \
+	iana-if-type.fxs ietf-interfaces.fxs $(PCG_DIR)/ietf_interfaces_ns.py \
+	$(PCG_DIR)/ietf_netconf_monitoring_ns.py \
+	$(CDB_DIR) ssh-keydir init_interfaces.xml
+	@echo "Build complete"
+
+$(PCG_DIR)/ietf_netconf_monitoring_ns.py: $(CONFD_DIR)/etc/confd/ietf-netconf-monitoring.fxs
```
Comment (cohult) [inline confdgnmi/Makefile]: Since they seem to not be used, I suggest skipping generating *_ns.py. It is old school anyway. Cool kids use the CsNode ConfD schema.

File: confdgnmi/README.md [hunk 19]
```diff
@@ -0,0 +1,283 @@
+Version history
+
+```
+--------------------------------------------------------------------------------
+Version:  Date:         Description:                        Author:
```
Comment (cohult) [inline confdgnmi/README.md]: Don't see the need for version numbering

File: confdgnmi/README.md [hunk 20]
```diff
@@ -0,0 +1,283 @@
...
+    * `updates_only` not supported
+    * `heartbeat_interval` not supported
+    * `sync_response` not generated
+* all values `TypedValues` are used as strings (`string_val`)
+* gNMI Path is converted to XPath or formatted path with simple string operations 
+  no datamodel knowledge used) 
+  
+## Investigations/TODOs
+
+TODO converting XPath to ConfD formatted path/keypath and back.
```
Comment (cohult) [inline confdgnmi/README.md:124]: I would primarily use RESTCONF instead of MAAPI or NETCONF

File: confdgnmi/README.md [hunk 21]
```diff
@@ -0,0 +1,283 @@
...
+`Capability` returns list of `supported_models` and list of `encodings`.
+It can also return list of gNMI extensions.  
+
+Each model has following attributes: 
+ 
+`name` (string), `organization` (string), `version`
+
+##### Implementation 
+
+Capability information can be fetched from `"/ncm:netconf-state/ncm:capabilities/ncm:capability"`
```
Comment (cohult) [inline confdgnmi/README.md:146]: Perhaps using RESTCONF "GET /restconf/data/ietf-yang-library:modules-state"?

File: confdgnmi/README.md [hunk 22]
```diff
@@ -0,0 +1,283 @@
...
+    const char *revision;
+    const char *module;
+}; 
+```
+
+`name` -> `module`
+`organization` -> TODO (cannot get with ConfD API, return empty string)
+`version` -> `revision` (return revision string -date ?)
+
+TODO: Pass yang file content as extension? HOW to get yang files from ConfD?
```
Comment (cohult) [inline confdgnmi/README.md:166]: Perhaps using RESTCONF  "GET /restconf/tailf/modules/dhcp/2019-02-14"

File: confdgnmi/README.md [hunk 23]
```diff
@@ -0,0 +1,283 @@
...
+`update`, each with value.  
+
+* Since there is a path associated with each value, for requests on nested lists
+we have to find out, which elements are keys and add them to the Path(s).
+
+* What value type should be for empty or presence container ??? (bool ???)
+
+##### Implementation 
+
+`maapi_save_config`  - can be used even for operational data (` MAAPI_CONFIG_WITH_OPER`).
```
Comment (cohult) [inline confdgnmi/README.md:194]: RESTCONF is another option where you can choose XML or JSON

File: confdgnmi/README.md (hunk 20 above)
Comment (micnovak) [reply confdgnmi/README.md:124]: Yes, RESTCONF adapter would be another option (or even MIXED adapter combining approaches). RESTCONF requires ConfD Premium.

File: confdgnmi/Makefile (hunk 17 above)
Comment (micnovak) [reply confdgnmi/Makefile:1]: resolved

File: confdgnmi/Makefile (hunk 18 above)
Comment (micnovak) [reply confdgnmi/Makefile]: _ns.py removed

File: confdgnmi/src/confd_gnmi_api_adapter.py (hunk 14 above)
Comment (micnovak) [reply confdgnmi/src/confd_gnmi_api_adapter.py]: Removed

File: confdgnmi/src/confd_gnmi_api_adapter.py (hunk 15 above)
Comment (micnovak) [reply confdgnmi/src/confd_gnmi_api_adapter.py:212]: ` [self.get_xml_leaves(e) for e in elem];` vreates array in array, in array .... . I need only array

File: confdgnmi/README.md (hunk 19 above)
Comment (micnovak) [reply confdgnmi/README.md]: removed

File: confdgnmi/README.md (hunk 20 above)
Comment (micnovak) [reply confdgnmi/README.md:124]: RESTCONF can be added in next milestones

File: confdgnmi/README.md (hunk 21 above)
Comment (micnovak) [reply confdgnmi/README.md:146]: RESTCONF can be added in next milestones

File: confdgnmi/README.md (hunk 22 above)
Comment (micnovak) [reply confdgnmi/README.md:166]: RESTCONF can be added in next milestones

File: confdgnmi/README.md (hunk 23 above)
Comment (micnovak) [reply confdgnmi/README.md:194]: RESTCONF can be added in next milestones

**Comment by jojohans** [comment]
Michal, can you sign off the commit(s)?
